Crash when compiling async let body with if expression

### Description

The following produces a crash on godbolt's nightly build, and confusing diagnostics in Xcode 16 beta 3

### Reproduction

```swift
func f() async {
    let s1: String? = ""
    async let val = if let s2 = s1 {
        g(s2)
    } else {
        0
    }

    await print(val)
}

func g(_: String) async -> Int {
    return 1
}
```
In Xcode, we get the following diagnostic:
```
% swiftc test.swift
test.swift:3:21: error: closure captures 's2' before it is declared
 1 | func f() async {
 2 |     let s1: String? = ""
 3 |     async let val = if let s2 = s1 {
   |                     |      `- note: captured value declared here
   |                     `- error: closure captures 's2' before it is declared
 4 |         g(s2)
   |           `- note: captured here
 5 |     } else {
 6 |         0
 ```

### Stack dump

```text
Godbolt doesn't produced a symbolicated stack, but in case it's useful:

swift-frontend: /home/build-user/swift/lib/SILGen/SILGenDecl.cpp:914: void (anonymous namespace)::LetValueInitialization::bindValue(swift::SILValue, swift::Lowering::SILGenFunction &, bool, swift::SILLocation): Assertion `!SGF.VarLocs.count(vd) && "Already emitted this vardecl?"' failed.
Please submit a bug report (https://swift.org/contributing/#reporting-bugs) and include the crash backtrace.
Stack dump:
0.	Program arguments: /opt/compiler-explorer/swift-nightly/usr/bin/swift-frontend -frontend -S -primary-file <source> -target x86_64-unknown-linux-gnu -disable-objc-interop -g -debug-info-format=dwarf -dwarf-version=4 -swift-version 6 -empty-abi-descriptor -Xcc -working-directory -Xcc /app -resource-dir /opt/compiler-explorer/swift-nightly/usr/lib/swift -enable-anonymous-context-mangled-names -file-compilation-dir /app -Xllvm --x86-asm-syntax=intel -module-name output -in-process-plugin-server-path /opt/compiler-explorer/swift-nightly/usr/lib/swift/host/libSwiftInProcPluginServer.so -plugin-path /opt/compiler-explorer/swift-nightly/usr/lib/swift/host/plugins -plugin-path /opt/compiler-explorer/swift-nightly/usr/local/lib/swift/host/plugins -o /app/output.s
1.	Swift version 6.0-dev (LLVM 0ad8ad0245d47b4, Swift 01bd2b4f8c84d45)
2.	Compiling with the current language version
3.	While evaluating request ASTLoweringRequest(Lowering AST to SIL for file "<source>")
4.	While silgen emitFunction SIL function "@$s6output1fyyYaF".
 for 'f()' (at <source>:1:1)
5.	While silgen closureexpr SIL function "@$s6output1fyyYaFSiyYacfu_".
 for expression at [<source>:3:21 - line:7:5] RangeText="if let s2 = s1 {
        g(s2)
    } else {
        0
    "
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  swift-frontend 0x00005ccac74856b7
1  swift-frontend 0x00005ccac74833be
2  swift-frontend 0x00005ccac7485d2f
3  libc.so.6      0x000078fbb3442520
4  libc.so.6      0x000078fbb34969fc pthread_kill + 300
5  libc.so.6      0x000078fbb3442476 raise + 22
6  libc.so.6      0x000078fbb34287f3 abort + 211
7  libc.so.6      0x000078fbb342871b
8  libc.so.6      0x000078fbb3439e96
9  swift-frontend 0x00005ccac0f6f62c
10 swift-frontend 0x00005ccac0f6f0b2
11 swift-frontend 0x00005ccac0f71cd7
12 swift-frontend 0x00005ccac0f673b6
13 swift-frontend 0x00005ccac0f71289
14 swift-frontend 0x00005ccac0f71f20
15 swift-frontend 0x00005ccac0ef2359
16 swift-frontend 0x00005ccac0ef20fb
17 swift-frontend 0x00005ccac0f81e72
18 swift-frontend 0x00005ccac0f6b8a6
19 swift-frontend 0x00005ccac1038a98
20 swift-frontend 0x00005ccac103630d
21 swift-frontend 0x00005ccac0fa3c4b
22 swift-frontend 0x00005ccac0f914f4
23 swift-frontend 0x00005ccac0f81f46
24 swift-frontend 0x00005ccac103af19
25 swift-frontend 0x00005ccac0fb44b3
26 swift-frontend 0x00005ccac0edce84
27 swift-frontend 0x00005ccac0ee6bd6
28 swift-frontend 0x00005ccac19b48af
29 swift-frontend 0x00005ccac0ee0185
30 swift-frontend 0x00005ccac0fa5f02
31 swift-frontend 0x00005ccac0fa5c75
32 swift-frontend 0x00005ccac0f9106b
33 swift-frontend 0x00005ccac0f8524c
34 swift-frontend 0x00005ccac0f61432
35 swift-frontend 0x00005ccac0f6150a
36 swift-frontend 0x00005ccac0f0fff9
37 swift-frontend 0x00005ccac0f6a8f2
38 swift-frontend 0x00005ccac0f72722
39 swift-frontend 0x00005ccac1037761
40 swift-frontend 0x00005ccac103630d
41 swift-frontend 0x00005ccac0fb413f
42 swift-frontend 0x00005ccac0eddb22
43 swift-frontend 0x00005ccac0edeee2
44 swift-frontend 0x00005ccac0edbf6c
45 swift-frontend 0x00005ccac0ee2190
46 swift-frontend 0x00005ccac1035c34
47 swift-frontend 0x00005ccac0ee7b15
48 swift-frontend 0x00005ccac0ee302c
49 swift-frontend 0x00005ccac0729e1f
50 swift-frontend 0x00005ccac073b009
51 swift-frontend 0x00005ccac072d1e1
52 swift-frontend 0x00005ccac072bf01
53 swift-frontend 0x00005ccac0550754
54 libc.so.6      0x000078fbb3429d90
55 libc.so.6      0x000078fbb3429e40 __libc_start_main + 128
56 swift-frontend 0x00005ccac054f7ee

*** Signal 11: Backtracing from 0x78fbb3428898... done ***

*** Program crashed: Bad pointer dereference at 0x0000000000000000 ***

Thread 0 "swift-frontend" crashed:

0  0x000078fbb3428898 <unknown> in libc.so.6


Registers:

rax 0x0000000000000000  0
rdx 0x000078fbb3a7ed00  00 ed a7 b3 fb 78 00 00 10 f7 a7 b3 fb 78 00 00  ·í§³ûx···÷§³ûx··
rcx 0x000078fbb34969fc  41 89 c5 41 f7 dd 3d 00 f0 ff ff b8 00 00 00 00  A·ÅA÷Ý=·ðÿÿ¸····
rbx 0x0000000000000006  6
rsi 0x0000000000000001  1
rdi 0x0000000000000001  1
rbp 0x000078fbb361be90  01 00 00 00 01 00 00 00 00 ed a7 b3 fb 78 00 00  ·········í§³ûx··
rsp 0x00007ffd43ba2b70  20 00 00 00 00 00 00 00 a0 b6 61 b3 fb 78 00 00   ······· ¶a³ûx··
 r8 0x0000000000000000  0
 r9 0x0000000000000000  0
r10 0x0000000000000008  8
r11 0x0000000000000246  582
r12 0x00005ccac9106469  2f 68 6f 6d 65 2f 62 75 69 6c 64 2d 75 73 65 72  /home/build-user
r13 0x0000000000000392  914
r14 0x00005ccac91dfb94  21 53 47 46 2e 56 61 72 4c 6f 63 73 2e 63 6f 75  !SGF.VarLocs.cou
r15 0x00005ccacb5d4450  88 b1 57 c7 ca 5c 00 00 01 00 00 00 00 00 00 00  ·±WÇÊ\··········
rip 0x000078fbb3428898  f4 83 3d 00 36 1f 00 05 75 14 c7 05 f4 35 1f 00  ô·=·6···u·Ç·ô5··

rflags 0x0000000000010246  ZF PF

cs 0x0033  fs 0x0000  gs 0x0000


Images (27 omitted):

0x000078fbb3400000–0x000078fbb35bc341 490fef8403240c91833978d494d39e537409b92e libc.so.6 /lib/x86_64-linux-gnu/libc.so.6

Backtrace took 0.01s

Program terminated with signal: SIGSEGV
Compiler returned: 139
```
```


### Expected behavior

Successfully compiles

### Environment

Godbolt:
```
Swift version 6.0-dev (LLVM 0ad8ad0245d47b4, Swift 01bd2b4f8c84d45)
Target: x86_64-unknown-linux-gnu
```
Xcode:
```
swift-driver version: 1.111.2 Apple Swift version 6.0 (swiftlang-6.0.0.5.15 clang-1600.0.22.6)
Target: arm64-apple-macosx14.0
```

### Additional information

_No response_

Another variant:
```swift
import Foundation

func sample(arg: Int?) async {
  // error: closure captures 'arg' before it is declared
  async let exampleOne = if let arg { arg } else { 0 }
  async let exampleTwo = switch arg {
    // error: closure captures 'value' before it is declared
    case .some(let value): value
    case .none: 0
  }
}
```
Getting errors:
```log
a.swift:4:26: error: closure captures 'arg' before it is declared
 2 |
 3 | func sample(arg: Int?) async {
 4 |   async let exampleOne = if let arg { arg } else { 0 }
   |                          |      |     `- note: captured here
   |                          |      `- note: captured value declared here
   |                          `- error: closure captures 'arg' before it is declared
 5 |   async let exampleTwo = switch arg {
 6 |     case .some(let value): value

a.swift:5:26: error: closure captures 'value' before it is declared
 3 | func sample(arg: Int?) async {
 4 |   async let exampleOne = if let arg { arg } else { 0 }
 5 |   async let exampleTwo = switch arg {
   |                          `- error: closure captures 'value' before it is declared
 6 |     case .some(let value): value
   |                    |       `- note: captured here
   |                    `- note: captured value declared here
 7 |     case .none: 0
 8 |   }
 ```

--- a/include/swift/AST/DeclContext.h
+++ b/include/swift/AST/DeclContext.h
@@ -261,7 +261,8 @@ class alignas(1 << DeclContextAlignInBits) DeclContext
   friend class Initializer; // uses setParent
   friend class AutoClosureExpr; // uses setParent
   friend class AbstractClosureExpr; // uses setParent
-  
+  friend class Decl; // uses setParent
+
   template<class A, class B, class C>
   friend struct ::llvm::CastInfo;
   

--- a/lib/AST/Decl.cpp
+++ b/lib/AST/Decl.cpp
@@ -767,6 +767,9 @@ void Decl::setInvalid() {
 
 void Decl::setDeclContext(DeclContext *DC) { 
   Context = DC;
+  // If this Decl is also a DeclContext, we need to update its parent too.
+  if (auto *thisDC = dyn_cast<DeclContext>(this))
+    thisDC->setParent(DC);
 }
 
 bool Decl::isUserAccessible() const {

--- a/lib/Sema/TypeCheckCaptures.cpp
+++ b/lib/Sema/TypeCheckCaptures.cpp
@@ -241,10 +241,9 @@ public:
   }
 
   LazyInitializerWalking getLazyInitializerWalkingBehavior() override {
-    // We don't want to walk into lazy initializers because they're not
-    // really present at this level.  We'll catch them when processing
-    // the getter.
-    return LazyInitializerWalking::None;
+    // Captures for lazy initializers are computed as part of the parent
+    // accessor.
+    return LazyInitializerWalking::InAccessor;
   }
 
   MacroWalking getMacroWalkingBehavior() const override {
@@ -642,12 +641,6 @@ public:
 
     if (auto *PEE = dyn_cast<PackElementExpr>(E))
       return walkToPackElementExpr(PEE);
-
-    // Look into lazy initializers.
-    if (auto *LIE = dyn_cast<LazyInitializerExpr>(E)) {
-      LIE->getSubExpr()->walk(*this);
-      return Action::Continue(E);
-    }
 
     // When we see a reference to the 'super' expression, capture 'self' decl.
     if (auto *superE = dyn_cast<SuperRefExpr>(E)) {

--- a/lib/Sema/TypeCheckConcurrency.cpp
+++ b/lib/Sema/TypeCheckConcurrency.cpp
@@ -2926,6 +2926,13 @@ namespace {
       return MacroWalking::Expansion;
     }
 
+    LazyInitializerWalking getLazyInitializerWalkingBehavior() override {
+      // We want to walk lazy initializers as part of their implicit getters
+      // since we're interested in querying capture information, and captures
+      // for lazy inits are computed as part of type-checking the accessor.
+      return LazyInitializerWalking::InAccessor;
+    }
+
     PreWalkResult<Pattern *> walkToPatternPre(Pattern *pattern) override {
       // Walking into patterns leads to nothing good because then we
       // end up visiting the AccessorDecls of a top-level

--- a/lib/Sema/TypeCheckDeclPrimary.cpp
+++ b/lib/Sema/TypeCheckDeclPrimary.cpp
@@ -1337,7 +1337,7 @@ Expr *DefaultArgumentExprRequest::evaluate(Evaluator &evaluator,
 
   // Walk the checked initializer and contextualize any closures
   // we saw there.
-  TypeChecker::contextualizeInitializer(dc, initExpr);
+  TypeChecker::contextualizeExpr(initExpr, dc);
   TypeChecker::checkInitializerEffects(dc, initExpr);
 
   return initExpr;

--- a/lib/Sema/TypeCheckExpr.cpp
+++ b/lib/Sema/TypeCheckExpr.cpp
@@ -772,7 +772,7 @@ Expr *CallerSideDefaultArgExprRequest::evaluate(
     return new (ctx) ErrorExpr(initExpr->getSourceRange(), paramTy);
   }
   if (param->getDefaultArgumentKind() == DefaultArgumentKind::ExpressionMacro) {
-    TypeChecker::contextualizeCallSideDefaultArgument(dc, initExpr);
+    TypeChecker::contextualizeExpr(initExpr, dc);
     TypeChecker::checkCallerSideDefaultArgumentEffects(dc, initExpr);
   }
   return initExpr;

--- a/lib/Sema/TypeCheckStmt.cpp
+++ b/lib/Sema/TypeCheckStmt.cpp
@@ -66,8 +66,13 @@ namespace {
   /// DeclContext.
   class ContextualizeClosuresAndMacros : public ASTWalker {
     DeclContext *ParentDC;
-  public:
+
     ContextualizeClosuresAndMacros(DeclContext *parent) : ParentDC(parent) {}
+
+  public:
+    static void contextualize(ASTNode node, DeclContext *DC) {
+      node.walk(ContextualizeClosuresAndMacros(DC));
+    }
 
     MacroWalking getMacroWalkingBehavior() const override {
       return MacroWalking::ArgumentsAndExpansion;
@@ -103,7 +108,7 @@ namespace {
 
       if (auto CE = dyn_cast<ClosureExpr>(E)) {
         CE->setParent(ParentDC);
-        CE->getBody()->walk(ContextualizeClosuresAndMacros(CE));
+        contextualize(CE->getBody(), CE);
 
         TypeChecker::computeCaptures(CE);
         return Action::SkipNode(E);
@@ -196,21 +201,13 @@ namespace {
   };
 } // end anonymous namespace
 
-void TypeChecker::contextualizeInitializer(Initializer *DC, Expr *E) {
-  ContextualizeClosuresAndMacros CC(DC);
-  E->walk(CC);
-}
-
-void TypeChecker::contextualizeCallSideDefaultArgument(DeclContext *DC,
-                                                       Expr *E) {
-  ContextualizeClosuresAndMacros CC(DC);
-  E->walk(CC);
+void TypeChecker::contextualizeExpr(Expr *E, DeclContext *DC) {
+  ContextualizeClosuresAndMacros::contextualize(E, DC);
 }
 
 void TypeChecker::contextualizeTopLevelCode(TopLevelCodeDecl *TLCD) {
-  ContextualizeClosuresAndMacros CC(TLCD);
   if (auto *body = TLCD->getBody())
-    body->walk(CC);
+    ContextualizeClosuresAndMacros::contextualize(body, TLCD);
 }
 
 namespace {
@@ -1033,7 +1030,7 @@ public:
   /// Type-check an entire function body.
   bool typeCheckBody(BraceStmt *&S) {
     bool HadError = typeCheckStmt(S);
-    S->walk(ContextualizeClosuresAndMacros(DC));
+    ContextualizeClosuresAndMacros::contextualize(S, DC);
     return HadError;
   }
 
@@ -2915,7 +2912,7 @@ TypeCheckFunctionBodyRequest::evaluate(Evaluator &eval,
         body = *optBody;
         alreadyTypeChecked = true;
 
-        body->walk(ContextualizeClosuresAndMacros(AFD));
+        ContextualizeClosuresAndMacros::contextualize(body, AFD);
       }
     }
   }

--- a/lib/Sema/TypeCheckStmt.cpp
+++ b/lib/Sema/TypeCheckStmt.cpp
@@ -61,17 +61,18 @@ using namespace swift;
 #define DEBUG_TYPE "TypeCheckStmt"
 
 namespace {
-  /// After forming autoclosures, we must re-parent any closure expressions
-  /// nested inside the autoclosure, because the autoclosure introduces a new
-  /// DeclContext.
-  class ContextualizeClosuresAndMacros : public ASTWalker {
+  /// After forming autoclosures and lazy initializer getters, we must update
+  /// the DeclContexts for any AST nodes that store the DeclContext they're
+  /// within. This includes e.g closures and decls, as well as some other
+  /// expressions, statements, and patterns.
+  class ContextualizationWalker : public ASTWalker {
     DeclContext *ParentDC;
 
-    ContextualizeClosuresAndMacros(DeclContext *parent) : ParentDC(parent) {}
+    ContextualizationWalker(DeclContext *parent) : ParentDC(parent) {}
 
   public:
     static void contextualize(ASTNode node, DeclContext *DC) {
-      node.walk(ContextualizeClosuresAndMacros(DC));
+      node.walk(ContextualizationWalker(DC));
     }
 
     MacroWalking getMacroWalkingBehavior() const override {
@@ -85,34 +86,7 @@ namespace {
     }
 
     PreWalkResult<Expr *> walkToExprPre(Expr *E) override {
-      if (auto CE = dyn_cast<AutoClosureExpr>(E)) {
-        CE->setParent(ParentDC);
-
-        // Recurse into the autoclosure body with the new ParentDC.
-        auto oldParentDC = ParentDC;
-        ParentDC = CE;
-        CE->getBody()->walk(*this);
-        ParentDC = oldParentDC;
-
-        TypeChecker::computeCaptures(CE);
-        return Action::SkipNode(E);
-      } 
-
-      if (auto CapE = dyn_cast<CaptureListExpr>(E)) {
-        // Capture lists need to be reparented to enclosing autoclosures
-        // and/or initializers of property wrapper backing properties
-        // (because they subsume initializers associated with wrapped
-        // properties).
-        if (isa<AutoClosureExpr>(ParentDC) ||
-            isPropertyWrapperBackingPropertyInitContext(ParentDC)) {
-          for (auto &Cap : CapE->getCaptureList()) {
-            Cap.PBD->setDeclContext(ParentDC);
-            Cap.getVar()->setDeclContext(ParentDC);
-          }
-        }
-      }
-
-      if (auto CE = dyn_cast<ClosureExpr>(E)) {
+      if (auto *CE = dyn_cast<AbstractClosureExpr>(E)) {
         CE->setParent(ParentDC);
         contextualize(CE->getBody(), CE);
 
@@ -121,12 +95,13 @@ namespace {
       }
 
       // Caller-side default arguments need their @autoclosures checked.
-      if (auto *DAE = dyn_cast<DefaultArgumentExpr>(E))
+      if (auto *DAE = dyn_cast<DefaultArgumentExpr>(E)) {
         if (DAE->isCallerSide() &&
             (DAE->getParamDecl()->isAutoClosure() ||
              (DAE->getParamDecl()->getDefaultArgumentKind() ==
               DefaultArgumentKind::ExpressionMacro)))
           DAE->getCallerSideDefaultExpr()->walk(*this);
+      }
 
       // Macro expansion expressions require a DeclContext as well.
       if (auto macroExpansion = dyn_cast<MacroExpansionExpr>(E)) {
@@ -136,26 +111,59 @@ namespace {
       return Action::Continue(E);
     }
 
-    /// We don't want to recurse into most local declarations.
-    PreWalkAction walkToDeclPre(Decl *D) override {
-      // But we do want to walk into the initializers of local
-      // variables.
-      return Action::VisitNodeIf(isa<PatternBindingDecl>(D));
+    PreWalkResult<Pattern *> walkToPatternPre(Pattern *P) override {
+      // A couple of patterns store DeclContexts.
+      if (auto *EP = dyn_cast<ExprPattern>(P))
+        EP->setDeclContext(ParentDC);
+      if (auto *EP = dyn_cast<EnumElementPattern>(P))
+        EP->setDeclContext(ParentDC);
+
+      return Action::Continue(P);
     }
 
-  private:
-    static bool isPropertyWrapperBackingPropertyInitContext(DeclContext *DC) {
-      auto *init = dyn_cast<PatternBindingInitializer>(DC);
-      if (!init)
-        return false;
+    PreWalkResult<Stmt *> walkToStmtPre(Stmt *S) override {
+      // The ASTWalker doesn't walk the case body variables, contextualize them
+      // ourselves.
+      if (auto *CS = dyn_cast<CaseStmt>(S)) {
+        for (auto *CaseVar : CS->getCaseBodyVariablesOrEmptyArray())
+          CaseVar->setDeclContext(ParentDC);
+      }
+      // A few statements store DeclContexts, update them.
+      if (auto *BS = dyn_cast<BreakStmt>(S))
+        BS->setDeclContext(ParentDC);
+      if (auto *CS = dyn_cast<ContinueStmt>(S))
+        CS->setDeclContext(ParentDC);
+      if (auto *FS = dyn_cast<FallthroughStmt>(S))
+        FS->setDeclContext(ParentDC);
 
-      if (auto *PB = init->getBinding()) {
-        auto *var = PB->getSingleVar();
-        return var && var->getOriginalWrappedProperty(
-                          PropertyWrapperSynthesizedPropertyKind::Backing);
+      return Action::Continue(S);
+    }
+
+    PreWalkAction walkToDeclPre(Decl *D) override {
+      // We may encounter some decls parented outside of a local context, e.g
+      // VarDecls in TopLevelCodeDecls are parented to the file. In such cases,
+      // assume the DeclContext they already have is correct, autoclosures
+      // and lazy var inits cannot be defined in such contexts anyway.
+      if (!D->getDeclContext()->isLocalContext())
+        return Action::SkipNode();
+
+      D->setDeclContext(ParentDC);
+
+      // Auxiliary decls need to have their contexts adjusted too.
+      if (auto *VD = dyn_cast<VarDecl>(D)) {
+        VD->visitAuxiliaryDecls([&](VarDecl *D) {
+          D->setDeclContext(ParentDC);
+        });
       }
 
-      return false;
+      // We don't currently support peer macro declarations in local contexts,
+      // however we don't reject them either; so just to be safe, adjust their
+      // context too.
+      D->visitAuxiliaryDecls([&](Decl *D) {
+        D->setDeclContext(ParentDC);
+      });
+
+      return Action::VisitNodeIf(isa<PatternBindingDecl>(D));
     }
   };
 
@@ -208,12 +216,12 @@ namespace {
 } // end anonymous namespace
 
 void TypeChecker::contextualizeExpr(Expr *E, DeclContext *DC) {
-  ContextualizeClosuresAndMacros::contextualize(E, DC);
+  ContextualizationWalker::contextualize(E, DC);
 }
 
 void TypeChecker::contextualizeTopLevelCode(TopLevelCodeDecl *TLCD) {
   if (auto *body = TLCD->getBody())
-    ContextualizeClosuresAndMacros::contextualize(body, TLCD);
+    ContextualizationWalker::contextualize(body, TLCD);
 }
 
 namespace {
@@ -1036,7 +1044,7 @@ public:
   /// Type-check an entire function body.
   bool typeCheckBody(BraceStmt *&S) {
     bool HadError = typeCheckStmt(S);
-    ContextualizeClosuresAndMacros::contextualize(S, DC);
+    ContextualizationWalker::contextualize(S, DC);
     return HadError;
   }
 
@@ -2918,7 +2926,7 @@ TypeCheckFunctionBodyRequest::evaluate(Evaluator &eval,
         body = *optBody;
         alreadyTypeChecked = true;
 
-        ContextualizeClosuresAndMacros::contextualize(body, AFD);
+        ContextualizationWalker::contextualize(body, AFD);
       }
     }
   }

--- a/lib/Sema/TypeCheckStmt.cpp
+++ b/lib/Sema/TypeCheckStmt.cpp
@@ -163,7 +163,10 @@ namespace {
         D->setDeclContext(ParentDC);
       });
 
-      return Action::VisitNodeIf(isa<PatternBindingDecl>(D));
+      // Only recurse into decls that aren't themselves DeclContexts. This
+      // allows us to visit e.g initializers for PatternBindingDecls and
+      // accessors for VarDecls.
+      return Action::SkipNodeIf(isa<DeclContext>(D));
     }
   };
 

--- a/lib/Sema/TypeCheckStmt.cpp
+++ b/lib/Sema/TypeCheckStmt.cpp
@@ -78,6 +78,12 @@ namespace {
       return MacroWalking::ArgumentsAndExpansion;
     }
 
+    LazyInitializerWalking getLazyInitializerWalkingBehavior() override {
+      // Don't walk lazy initializers, we contextualize the getter body
+      // specially when synthesizing.
+      return LazyInitializerWalking::None;
+    }
+
     PreWalkResult<Expr *> walkToExprPre(Expr *E) override {
       if (auto CE = dyn_cast<AutoClosureExpr>(E)) {
         CE->setParent(ParentDC);

--- a/lib/Sema/TypeCheckStorage.cpp
+++ b/lib/Sema/TypeCheckStorage.cpp
@@ -618,7 +618,7 @@ static void checkAndContextualizePatternBindingInit(PatternBindingDecl *binding,
 
   if (auto *initContext = binding->getInitContext(i)) {
     auto *init = binding->getInit(i);
-    TypeChecker::contextualizeInitializer(initContext, init);
+    TypeChecker::contextualizeExpr(init, initContext);
   }
 }
 
@@ -3200,7 +3200,7 @@ static void typeCheckSynthesizedWrapperInitializer(VarDecl *wrappedVar,
   // will be done in visitPatternBindingDecl.
   if (!contextualize)
     return;
-  TypeChecker::contextualizeInitializer(initContext, initializer);
+  TypeChecker::contextualizeExpr(initializer, initContext);
   checkPropertyWrapperActorIsolation(wrappedVar, initializer);
   TypeChecker::checkInitializerEffects(initContext, initializer);
 }

--- a/lib/Sema/TypeCheckStorage.cpp
+++ b/lib/Sema/TypeCheckStorage.cpp
@@ -1679,6 +1679,7 @@ namespace {
       // If we find a closure, update its declcontext and do *not* walk into it.
       if (auto CE = dyn_cast<AbstractClosureExpr>(E)) {
         CE->setParent(NewDC);
+        TypeChecker::computeCaptures(CE);
         return Action::SkipNode(E);
       }
 

--- a/lib/Sema/TypeCheckStorage.cpp
+++ b/lib/Sema/TypeCheckStorage.cpp
@@ -1660,76 +1660,6 @@ synthesizeRead2CoroutineGetterBody(AccessorDecl *getter, ASTContext &ctx) {
   return synthesizeTrivialGetterBody(getter, TargetImpl::Implementation, ctx);
 }
 
-namespace {
-  /// This ASTWalker explores an expression tree looking for expressions (which
-  /// are DeclContext's) and changes their parent DeclContext to NewDC.
-  /// TODO: We ought to consider merging this with
-  /// ContextualizeClosuresAndMacros, or better yet removing it in favor of
-  /// avoiding the recontextualization for lazy vars.
-  class RecontextualizeClosures : public ASTWalker {
-    DeclContext *NewDC;
-  public:
-    RecontextualizeClosures(DeclContext *NewDC) : NewDC(NewDC) {}
-
-    MacroWalking getMacroWalkingBehavior() const override {
-      return MacroWalking::ArgumentsAndExpansion;
-    }
-
-    PreWalkResult<Expr *> walkToExprPre(Expr *E) override {
-      // If we find a closure, update its declcontext and do *not* walk into it.
-      if (auto CE = dyn_cast<AbstractClosureExpr>(E)) {
-        CE->setParent(NewDC);
-        TypeChecker::computeCaptures(CE);
-        return Action::SkipNode(E);
-      }
-
-      return Action::Continue(E);
-    }
-
-    PreWalkResult<Pattern *> walkToPatternPre(Pattern *P) override {
-      if (auto *EP = dyn_cast<ExprPattern>(P))
-        EP->setDeclContext(NewDC);
-      if (auto *EP = dyn_cast<EnumElementPattern>(P))
-        EP->setDeclContext(NewDC);
-
-      return Action::Continue(P);
-    }
-
-    PreWalkResult<Stmt *> walkToStmtPre(Stmt *S) override {
-      // The ASTWalker doesn't walk the case body variables, contextualize them
-      // ourselves.
-      if (auto *CS = dyn_cast<CaseStmt>(S)) {
-        for (auto *CaseVar : CS->getCaseBodyVariablesOrEmptyArray())
-          CaseVar->setDeclContext(NewDC);
-      }
-      // A few statements store DeclContexts, update them.
-      if (auto *BS = dyn_cast<BreakStmt>(S))
-        BS->setDeclContext(NewDC);
-      if (auto *CS = dyn_cast<ContinueStmt>(S))
-        CS->setDeclContext(NewDC);
-      if (auto *FS = dyn_cast<FallthroughStmt>(S))
-        FS->setDeclContext(NewDC);
-
-      return Action::Continue(S);
-    }
-
-    PreWalkAction walkToDeclPre(Decl *D) override {
-      D->setDeclContext(NewDC);
-
-      // Auxiliary decls need to have their contexts adjusted too.
-      if (auto *VD = dyn_cast<VarDecl>(D)) {
-        VD->visitAuxiliaryDecls([&](VarDecl *D) {
-          D->setDeclContext(NewDC);
-        });
-      }
-
-      // Skip walking the children of any Decls that are also DeclContexts,
-      // they will already have the right parent.
-      return Action::SkipNodeIf(isa<DeclContext>(D));
-    }
-  };
-} // end anonymous namespace
-
 /// Synthesize the getter for a lazy property with the specified storage
 /// vardecl.
 static std::pair<BraceStmt *, bool>
@@ -1801,12 +1731,13 @@ synthesizeLazyGetterBody(AccessorDecl *Get, VarDecl *VD, VarDecl *Storage,
     InitValue = new (Ctx) ErrorExpr(SourceRange(), Tmp2VD->getTypeInContext());
   }
 
-  // Recontextualize any closure declcontexts nested in the initializer to
-  // realize that they are in the getter function.
+  // We may have stolen the self decl from a PatternBindingInitializer,
+  // re-parent it here.
   if (Get->hasImplicitSelfDecl())
     Get->getImplicitSelfDecl()->setDeclContext(Get);
 
-  InitValue->walk(RecontextualizeClosures(Get));
+  // Also re-contextualize any nodes nested in the initializer.
+  TypeChecker::contextualizeExpr(InitValue, Get);
 
   // Wrap the initializer in a LazyInitializerExpr to avoid walking it twice.
   auto initType = InitValue->getType();

--- a/lib/Sema/TypeChecker.h
+++ b/lib/Sema/TypeChecker.h
@@ -773,10 +773,9 @@ void computeCaptures(AbstractClosureExpr *ACE);
 /// Check for invalid captures from stored property initializers.
 void checkPatternBindingCaptures(IterableDeclContext *DC);
 
-/// Change the context of closures in the given initializer
-/// expression to the given context.
-void contextualizeInitializer(Initializer *DC, Expr *init);
-void contextualizeCallSideDefaultArgument(DeclContext *DC, Expr *init);
+/// Update the DeclContexts for AST nodes in a given DeclContext. This is
+/// necessary after type-checking since autoclosures may have been introduced.
+void contextualizeExpr(Expr *E, DeclContext *DC);
 void contextualizeTopLevelCode(TopLevelCodeDecl *TLCD);
 
 /// Retrieve the default type for the given protocol.

--- a/test/SILGen/if_expr.swift
+++ b/test/SILGen/if_expr.swift
@@ -659,6 +659,21 @@ struct LazyProp {
   }
 }
 
+// https://github.com/swiftlang/swift/issues/75294
+func testAsyncLet(_ x: Int?) async {
+  async let _ = if let i = x { i } else { 0 }
+  // CHECK-LABEL: sil private [ossa] @$s7if_expr12testAsyncLetyySiSgYaFSiyYaYbcfu_ : $@convention(thin) @Sendable @async @substituted <τ_0_0> (Optional<Int>) -> (@out τ_0_0, @error any Error) for <Int>
+
+  async let _ = if let i = x {
+    // CHECK-LABEL: sil private [ossa] @$s7if_expr12testAsyncLetyySiSgYaFSiyYaYbcfu0_ : $@convention(thin) @Sendable @async @substituted <τ_0_0> (Optional<Int>) -> (@out τ_0_0, @error any Error) for <Int>
+    lazy var y = i
+    // CHECK-LABEL: sil private [lazy_getter] [noinline] [ossa] @$s7if_expr12testAsyncLetyySiSgYaFSiyYaYbcfu0_1yL_Sivg : $@convention(thin) (@guaranteed { var Optional<Int> }, Int) -> Int
+    then y
+  } else {
+    0
+  }
+}
+
 func testNestedFallthrough1() throws -> Int {
   let x = if .random() {
     switch Bool.random() {

--- a/test/SILGen/skip_function_bodies_storage.swift
+++ b/test/SILGen/skip_function_bodies_storage.swift
@@ -39,3 +39,12 @@ extension S: P {}
 // CHECK: sil shared [serialized]{{.*}} @$s4main1SV11borrowedVarSivr : $@yield_once @convention(method) (S) -> @yields Int {
 // CHECK:   yield
 // CHECK: } // end sil function '$s4main1SV11borrowedVarSivr'
+
+// We type-check this function since it has a nested type, but we don't
+// type-check the implicit lazy getter for 'x'. As such, the nested autoclosure
+// won't have captures computed. Make sure we don't attempt to query the
+// captures.
+func testAutoclosureInLazyVar(_ y: Int?) {
+  struct R {}
+  lazy var x = y ?? 0
+}

--- a/test/SILGen/switch_expr.swift
+++ b/test/SILGen/switch_expr.swift
@@ -302,6 +302,12 @@ struct TestPropertyInit {
   }
 }
 
+// https://github.com/swiftlang/swift/issues/75294
+func testAsyncLet(_ x: Int?) async {
+  async let _ = switch x { case let i?: i default: 0 }
+  // CHECK-LABEL: sil private [ossa] @$s11switch_expr12testAsyncLetyySiSgYaFSiyYaYbcfu_ : $@convention(thin) @Sendable @async @substituted <τ_0_0> (Optional<Int>) -> (@out τ_0_0, @error any Error) for <Int>
+}
+
 func testNested(_ e: E) throws -> Int {
   switch e {
   case .a:

